adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1718,38 +1799,73 @@ private boolean stopCalled() {
         }
 
         /**
-         * Closes a connection to a client.
+         * Actively closes the connection to the device.
+         *
+         * @param reason The reason for closing the connection.
+         * @param sendDisconnectedEvent {@code true} if events shall be sent concerning the disconnect.
+         * @throws NullPointerException if reason is {@code null}.
+         */
+        protected final void close(final String reason, final boolean sendDisconnectedEvent) {
```

Comment:
shouldn't the whole class be final?

Reply:
For now I prefer to keep the individual methods marked as final here. When extracting `MqttDeviceEndpoint` to upper level (#3084) it might be convenient to keep the class non-final (for extension via the `AbstractVertxBasedMqttProtocolAdapter` sub classes).